internal/services/postgres/postgresql_flexible_server_virtual_endpoint_resource.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/hashicorp/go-azure-sdk/resource-manager/postgresql/2023-06-01-preview/virtualendpoints"
+	"github.com/hashicorp/go-azure-sdk/sdk/client/pollers"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/locks"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/postgres/custompollers"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type PostgresqlFlexibleServerVirtualEndpointResource struct{}
+
+type PostgresqlFlexibleServerVirtualEndpointModel struct {
+	Name            string `tfschema:"name"`
+	SourceServerId  string `tfschema:"source_server_id"`
+	ReplicaServerId string `tfschema:"replica_server_id"`
+	Type            string `tfschema:"type"`
+}
+
+var _ sdk.ResourceWithUpdate = PostgresqlFlexibleServerVirtualEndpointResource{}
+
+func (r PostgresqlFlexibleServerVirtualEndpointResource) ModelObject() interface{} {
+	return &PostgresqlFlexibleServerVirtualEndpointModel{}
+}
+
+func (r PostgresqlFlexibleServerVirtualEndpointResource) ResourceType() string {
+	return "azurerm_postgresql_flexible_server_virtual_endpoint"
+}
+
+func (r PostgresqlFlexibleServerVirtualEndpointResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return func(input interface{}, key string) (warnings []string, errors []error) {
+		v, ok := input.(string)
+		if !ok {
+			errors = append(errors, fmt.Errorf("expected %q to be a string", key))
+			return
+		}
+
+		if _, err := virtualendpoints.ParseVirtualEndpointID(v); err != nil {
+			errors = append(errors, err)
+		}
+
+		return
+	}
+}
+
+func (r PostgresqlFlexibleServerVirtualEndpointResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+func (r PostgresqlFlexibleServerVirtualEndpointResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:        pluginsdk.TypeString,
+			Description: "The name of the Virtual Endpoint",
+			ForceNew:    true,
+			Required:    true,
+		},
+		"source_server_id": {
+			Type:         pluginsdk.TypeString,
+			Description:  "The Resource ID of the *Source* Postgres Flexible Server this should be associated with",
+			ForceNew:     true,
+			Required:     true,
+			ValidateFunc: servers.ValidateFlexibleServerID,
+		},
+		"replica_server_id": {
+			Type:         pluginsdk.TypeString,
+			Description:  "The Resource ID of the *Replica* Postgres Flexible Server this should be associated with",
+			Required:     true,
+			ValidateFunc: servers.ValidateFlexibleServerID,
+		},
+		"type": {
+			Type:         pluginsdk.TypeString,
+			Description:  "The type of Virtual Endpoint",
+			ForceNew:     true,
+			Required:     true,
+			ValidateFunc: validation.StringInSlice(virtualendpoints.PossibleValuesForVirtualEndpointType(), true),
```

Comment:
The API only exposes one valid value for this at the moment `ReadWrite` even in the latest preview version of the API, but the [documentation](https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/concepts-read-replicas-virtual-endpoints) mentions two types of connection points, Read/Write and Read Only. 

It would be good to understand what the default behaviour of the API is e.g. is it possible to create this resource without specifying the `type` in the payload and if so what default value is returned.

The outcome of this would help determine whether we need to expose `type` at all or whether this can just be hardcoded to `ReadWrite` by the provider.

Reply:
> It would be good to understand what the default behaviour of the API is e.g. is it possible to create this resource without specifying the type in the payload and if so what default value is returned.

The `Type` argument is required, otherwise a 400 is returned.

The presumption is that Read replicas will become available at some point, and it would be as easy as updating the SDK to get access to the new types. 

I based this on the way the `virtualendpoints.VirtualEndpointType` is declared as an enum-like type, as well as how other providers such as AWS deal with replicas.

We could default it to `ReadWrite`, and that sort of satisfies brevity of declaration, as well as leaving the door open to future `virtualendpoints.VirtualEndpointType`s?

Reply:
Thanks for the clarification, I think this is fine as is with the background